File: cli/internal/client/llm.go
package client

import (
	"context"
	"net/url"
)

// ListLLMProviders returns the provider catalog with config status.
func (c *Client) ListLLMProviders(ctx context.Context) ([]map[string]any, error) {
	var resp struct {
		Providers []map[string]any `json:"providers"`
	}
	if err := c.GetJSON(ctx, "/api/llm/providers", &resp); err != nil {
		return nil, err
	}
	return resp.Providers, nil
}

// GetProviderModels returns the model catalog for a single provider.
func (c *Client) GetProviderModels(ctx context.Context, provider string) (map[string]any, error) {
	var out map[string]any
	if err := c.GetJSON(ctx, "/api/llm/providers/"+url.PathEscape(provider)+"/models", &out); err != nil {
		return nil, err
	}
	return out, nil
}

// ConfigureProvider writes provider config keys (api_key, auth_method, …).
// Each value is sent as-is; the server flags secret keys based on the catalog.
func (c *Client) ConfigureProvider(ctx context.Context, provider string, kv map[string]any) error {
	return c.PutJSON(ctx, "/api/llm/providers/"+url.PathEscape(provider), kv, nil)
}

// DeleteProviderConfig removes all stored config for a provider.
func (c *Client) DeleteProviderConfig(ctx context.Context, provider string) error {
	return c.Delete(ctx, "/api/llm/providers/"+url.PathEscape(provider)+"/config", nil)
}

// GetModelGroups returns current model_group + reasoning_effort assignments.
func (c *Client) GetModelGroups(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	if err := c.GetJSON(ctx, "/api/llm/model-groups", &out); err != nil {
		return nil, err
	}
	return out, nil
}

// UpdateModelGroups patches model_group / reasoning_efforts / default_provider.
func (c *Client) UpdateModelGroups(ctx context.Context, body map[string]any) error {
	return c.PutJSON(ctx, "/api/llm/model-groups", body, nil)
}

// DeviceCodeStartResponse is the GitHub device-code start payload.
type DeviceCodeStartResponse struct {
	VerificationURI string `json:"verification_uri"`
	UserCode        string `json:"user_code"`
	DeviceCode      string `json:"device_code"`
	ExpiresIn       int    `json:"expires_in"`
	Interval        int    `json:"interval"`
}

// DeviceCodeStart kicks off the GitHub Copilot device-code flow.
func (c *Client) DeviceCodeStart(ctx context.Context) (*DeviceCodeStartResponse, error) {
	var out DeviceCodeStartResponse
	if err := c.PostJSON(ctx, "/api/llm/auth/device-code/start", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeviceCodePollResponse mirrors the poll endpoint reply. Status is one of
// "pending", "expired", "complete", "error". When complete-and-unauthenticated,
// AccessToken is populated; otherwise the server stores the token under the
// active profile.
type DeviceCodePollResponse struct {
	Status      string `json:"status"`
	SlowDown    bool   `json:"slow_down,omitempty"`
	AccessToken string `json:"access_token,omitempty"`
	Error       string `json:"error,omitempty"`
}

// DeviceCodePoll polls GitHub for a token using a previously-issued device_code.
func (c *Client) DeviceCodePoll(ctx context.Context, deviceCode string) (*DeviceCodePollResponse, error) {
	var out DeviceCodePollResponse
	body := map[string]string{"device_code": deviceCode}
	if err := c.PostJSON(ctx, "/api/llm/auth/device-code/poll", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: app/cli/client/llm.py
"""LLM provider + model-group endpoints — `/api/llm/*`.

Mirrors `cli/internal/client/llm.go`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from urllib.parse import quote

from app.cli.client._base import Client


# ── providers ─────────────────────────────────────────────────────────────

async def list_llm_providers(client: Client) -> list[dict[str, Any]]:
    resp = await client.get_json("/api/llm/providers")
    if isinstance(resp, dict) and isinstance(resp.get("providers"), list):
        return [p for p in resp["providers"] if isinstance(p, dict)]
    return []


async def get_provider_models(
    client: Client,
    provider: str,
) -> dict[str, Any]:
    out = await client.get_json(f"/api/llm/providers/{quote(provider, safe='')}/models")
    return out if isinstance(out, dict) else {}


async def configure_provider(
    client: Client,
    provider: str,
    kv: dict[str, Any],
) -> None:
    await client.put_json(f"/api/llm/providers/{quote(provider, safe='')}", kv)


async def delete_provider_config(client: Client, provider: str) -> None:
    await client.delete(f"/api/llm/providers/{quote(provider, safe='')}/config")


# ── model groups ──────────────────────────────────────────────────────────

async def get_model_groups(client: Client) -> dict[str, Any]:
    out = await client.get_json("/api/llm/model-groups")
    return out if isinstance(out, dict) else {}


async def update_model_groups(
    client: Client,
    body: dict[str, Any],
) -> None:
    await client.put_json("/api/llm/model-groups", body)


# ── device-code flow ──────────────────────────────────────────────────────

@dataclass(frozen=True)
class DeviceCodeStart:
    verification_uri: str
    user_code: str
    device_code: str
    expires_in: int
    interval: int

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "DeviceCodeStart":
        return cls(
            verification_uri=str(d.get("verification_uri") or ""),
            user_code=str(d.get("user_code") or ""),
            device_code=str(d.get("device_code") or ""),
            expires_in=int(d.get("expires_in") or 0),
            interval=int(d.get("interval") or 5),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "verification_uri": self.verification_uri,
            "user_code": self.user_code,
            "device_code": self.device_code,
            "expires_in": self.expires_in,
            "interval": self.interval,
        }


@dataclass(frozen=True)
class DeviceCodePoll:
    status: str
    slow_down: bool
    access_token: str
    error: str

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "DeviceCodePoll":
        return cls(
            status=str(d.get("status") or ""),
            slow_down=bool(d.get("slow_down") or False),
            access_token=str(d.get("access_token") or ""),
            error=str(d.get("error") or ""),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "status": self.status,
            "slow_down": self.slow_down,
            "access_token": self.access_token,
            "error": self.error,
        }


async def device_code_start(client: Client) -> DeviceCodeStart:
    data = await client.post_json("/api/llm/auth/device-code/start")
    if not isinstance(data, dict):
        raise RuntimeError("unexpected device-code/start response")
    return DeviceCodeStart.from_dict(data)


async def device_code_poll(client: Client, device_code: str) -> DeviceCodePoll:
    data = await client.post_json(
        "/api/llm/auth/device-code/poll",
        {"device_code": device_code},
    )
    if not isinstance(data, dict):
        raise RuntimeError("unexpected device-code/poll response")
    return DeviceCodePoll.from_dict(data)
